### PR TITLE
state-surgery: genesis block generation

### PR DESCRIPTION
**Pull request by tynes**
**Description**

Generate the L2 genesis block given a hardhat and a deploy config.

### CODE REVIEW

File: state-surgery/genesis/helpers.go [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+		code := db.GetCode(addr)
+		if len(code) == 0 {
+			return fmt.Errorf("code not set for %s", name)
+		}
+	}
+	return nil
+}
+
+// AddressToCodeNamespace takes a predeploy address and computes
+// the address that
```
Comment (trianglesphere) [inline state-surgery/genesis/helpers.go:174]: dropped comment

**Review by trianglesphere — COMMENTED** [review]
2 small nits, but the rest of the code looks good

File: state-surgery/genesis/layer_two_test.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+	proxyAdmin, err := hh.GetDeployment("ProxyAdmin")
+	require.Nil(t, err)
+
+	for name, address := range predeploys.Predeploys {
+		addr := *address
+
+		account, ok := gen.Alloc[addr]
+		if !ok {
+			fmt.Println("not ok", name)
```
Comment (tynes) [inline state-surgery/genesis/layer_two_test.go:62]: delete this dead code

File: state-surgery/genesis/layer_one.go [hunk 3]
```diff
@@ -0,0 +1,22 @@
+package genesis
+
+import (
+	"github.com/ethereum-optimism/optimism/state-surgery/hardhat"
+	"github.com/ethereum-optimism/optimism/state-surgery/state"
+	"github.com/ethereum/go-ethereum/core"
+)
+
+// TODO(tynes): need bindings for all of the L1 contracts
```
Comment (protolambda) [inline state-surgery/genesis/layer_one.go:9]: Opened ENG-2659 to track this todo

**Review by protolambda — APPROVED** [review]
LGTM, resolve the nits and then good to go, and I opened an issue to track the new todo comment.

File: state-surgery/genesis/helpers.go [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+		return block, nil
+	} else if num, ok := tag.Number(); ok {
+		blockNumber := new(big.Int).SetInt64(num.Int64())
+		block, err := chain.BlockByNumber(context.Background(), blockNumber)
+		if err != nil {
+			return nil, err
+		}
+		return block, nil
+	} else {
+		return nil, errors.New("foo")
```
[the remark this answers is not in the document]
Comment (tynes) [reply state-surgery/genesis/helpers.go:231]: haha good catch, sorry about that, missed my self review

**Review by trianglesphere — APPROVED** [review]

**Pull request merged**